design/src/rfcs/rfc0012_constraint_traits.md

```diff
@@ -0,0 +1,424 @@
+[`aws_smithy_types::Blob`]: https://docs.rs/aws-smithy-types/latest/aws_smithy_types/struct.Blob.html
+[`aws_smithy_http_server::rejection::SmithyRejection`]: https://docs.rs/aws-smithy-http-server/latest/aws_smithy_http_server/rejection/enum.SmithyRejection.html
+[`aws_smithy_http_server::rejection::Deserialize`]: https://docs.rs/aws-smithy-http-server/latest/aws_smithy_http_server/rejection/struct.Deserialize.html
+[#1187]: https://github.com/awslabs/smithy-rs/issues/1187
+[`lazy_static!`](https://docs.rs/lazy_static/1.4.0/lazy_static/)
+
+Trait precedence and naming of the tuple struct
+-----------------------------------------------
+
+From [the spec]:
+
+> Some constraints can be applied to shapes as well as structure members. If a
+> constraint of the same type is applied to a structure member and the shape
+> that the member targets, the trait applied to the member takes precedence.
+
+```smithy
+structure ShoppingCart {
+    @range(min: 7, max:12)
+    numberOfItems: PositiveInteger
+}
+
+@range(min: 1)
+integer PositiveInteger
+```
+
+In the above example,
+
+> the range trait applied to numberOfItems takes precedence over the one
+> applied to PositiveInteger. The resolved minimum will be 7, and the maximum
+> 12.
+
+When the constraint trait is applied to a member shape, the tuple struct's name
+will be the PascalCased name of the member shape, `NumberOfItems`.
+
+[the spec]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html?highlight=enum#trait-precedence
+
+Unresolved questions
+--------------------
+
+1. Should we code-generate unsigned integer types (`u16`, `u32`, `u64`) when
+   the `range` trait is applied with `min` set to a value greater than or equal
+   to 0?
+    - A user has even suggested to use the `std::num::NonZeroUX` types (e.g.
+      [`NonZeroU64`]) when `range` is applied with `min` set to a value greater
+      than 0.
+2. In request deserialization, should we fail with the first violation and
+   immediately render a response, or attempt to parse the entire request and
+   provide a complete and structured report?
+3. Should we provide a mechanism for the service implementer to construct a
+   Rust type violating the modeled constraints in their business logic e.g. a
+   `T::new_unchecked()` constructor? This could be useful (1) when the user
+   _knows_ the provided inner value does not violate the constraints and
+   doesn't want to incur the performance penalty of the check; (2) when the
+   struct is in a transient invalid state. However:
+    - (2) is arguably a modelling mistake and a separate struct to represent
+      the transient state would be a better approach,
+    - the user can always use `unsafe` Rust to bypass the validation; and
+    - adding this constructor is a backwards-compatible change, so it can
+      always be added later if this feature is requested.
+
+[`NonZeroU64`]: https://doc.rust-lang.org/std/num/struct.NonZeroU64.html
+
+Alternative design
```

Comment:
I prefer the primary design, though with the addendum that I think we should provide conversions back to the primitive type. Something like `fn into_unconstrained`.

Reply:
Good point, need to add this to the doc. I was thinking of `into_inner` but `into_unconstrained` is more explicit.

Reply:
I also prefer the primary design as it fosters correctness.